common/src/IO/ObjParser.cpp

```diff
@@ -0,0 +1,166 @@
+                        try {
+                            tex = std::unique_ptr<Assets::Texture> (imageReader.readTexture(m_fs.openFile(IO::Path("textures/" + tokens[1] + ".png"))));
+                        } catch (const Exception& /*ex1*/) {
+                            try {
+                                tex = std::unique_ptr<Assets::Texture> (imageReader.readTexture(m_fs.openFile(IO::Path("textures/" + tokens[1] + ".jpg"))));
+                            } catch (const Exception& /*ex2*/) {
+                                current_material = 0;
+                                logger.warn() << "unable to find Neverball model material " << tokens[1];
+                            }
+                        }
+                        if (tex) {
+                            surface.addSkin(tex.release());
+                            last_material++;
+                            current_material = last_material;
+                        }
+                    } else if (tokens[0] == "f") {
+                        ObjFace face;
+                        face.m_material = current_material;
+                        for (size_t i = 1; i < tokens.size(); i++)
+                            face.m_vertices.push_back(ObjVertexRef(tokens[i]));
+                        faces.push_back(face);
+                    }
+                }
+            }
+            // Done parsing; setup bounds.
+            auto bounds = vm::bbox3f::builder();
+            if (positions.size() == 0) {
+                // passing empty bounds as bbox crashes the program, don't let it happen
+                throw ParserException("OBJ file has no vertices (so no valid bounding box)");
+            } else {
+                bounds.add(std::begin(positions), std::end(positions));
+            }
+            // Model construction prestart
+            // Begin model construction, part 1. Collation
+            size_t totalVertexCount = 0;
+            Renderer::TexturedIndexRangeMap::Size size;
+            for (const ObjFace & face : faces) {
+                size.inc(surface.skin(face.m_material), Renderer::PrimType::Polygon, face.m_vertices.size());
+                totalVertexCount += face.m_vertices.size();
+            }
+            // Model construction, part 2. Building
+            auto& frame = model->loadFrame(0, m_name, bounds.bounds());
+            Renderer::TexturedIndexRangeMapBuilder<Assets::EntityModelVertex::Type> builder(totalVertexCount, size);
+            // Actual build {
+            for (const ObjFace & face : faces) {
+                std::vector<Assets::EntityModelVertex> vertices;
+                for (const ObjVertexRef & ref : face.m_vertices) {
+                    size_t point = ref.m_position;
+                    if (point == 0)
+                        throw ParserException("OBJ file has vertex with no position (was this generated/parsed correctly?)");
+                    point--;
```

Comment:
`--point`

Reply:
Also please leave a comment as to why this needs decrementation.

Reply:
Resolved as of 19e8464609, including comment (summary: OBJ format indexes are 1-based)